bugfix: don't recompute full_hash or build_hash

tgamblin opened the pull request:
Fixes #19649.

There was an error introduced in #19209 where `full_hash()` and `build_hash()` are called on older specs that we've read in from the DB.

Since we evaluate hashes lazily, `node_dict_with_hashes()` has no way of knowing whether the spec just never had a `full_hash` or a `build_hash` or whether they need to be lazily computed. When we just had `dag_hash`, this wasn't an issue since it was always computed when concrete specs were stored -- it would always be there for specs read in from files.

To get around this ambiguity and to fix the issue, this makes two changes:

- [x] Always compute `dag_hash`, `build_hash`, and `full_hash` after concretization, so we know if one is present, then all should be, if they are avaialble.

- [x] Only store `full_hash` and `build_hash` if they are present. If not, don't bother recomputing.

This *may* make concretization slightly slower. If it does we could look at some other way to store whether full_hash and build_hash can be computed (e.g. we coudl store a `bool` on the `Spec` at concretize time), or we could look at other options like using JSON for these hashes to speed them up.  I think this solution is simpler for now.

tgamblin committed:
bugfix: don't recompute full_hash or build_hash

There was an error introduced in #19209 where `full_hash()` and
`build_hash()` are called on older specs that we've read in from the DB.

Since we evaluate hashes lazily, `node_dict_with_hashes()` has no way of
knowing whether the spec just never had a `full_hash` or a `build_hash`
or whether they need to be lazily computed. When we just had `dag_hash`,
this wasn't an issue since it was always computed when concrete specs
were stored -- it would always be there for specs read in from files.

To get around this ambiguity and to fix the issue, this makes two
changes:

- [x] Always compute `dag_hash`, `build_hash`, and `full_hash` after
      concretization, so we know if one is present, then all should be,
      if they are avaialble.

- [x] Only store `full_hash` and `build_hash` if they are present. If
      not, don't bother recomputing.

This *may* make concretization slightly slower. If it does we could look
at some other way to store whether full_hash and build_hash can be
computed (e.g. we coudl store a `bool` on the `Spec` at concretize time),
or we could look at other options like using JSON for these hashes to
speed them up.

tgamblin commented:
Not sure if this addresses @haampie's #19665 (pretty sure it doesn't yet)  but FYI @haampie.

haampie commented:
> Not sure if this addresses @haampie's #19665 (pretty sure it doesn't yet) but FYI @haampie.

Thanks. I guess this is something orthogonal indeed. The problem in #19665 is rather there's a mismatch between the defaults in `buildcache check` and `install` -- the former assumes `full_hash` is used by default, the latter only requires a `full_hash` on the binary cache if you pass a `--require-full-hash-match` flag.

adamjstewart commented:
> Fixes #19647

Did you mean #19649?

tgamblin commented:
> Did you mean #19649?

Yes!

scottwittenburg reviewed (commented):
This seems reasonable to me.  @scheibelp and I went back and forth a bit as to whether computing the full hash here was the right thing to do, and eventually we came to agreement that at this point, we wouldn't be here if the spec had been read from the database (and thus be recomputing the hash).  But we must have been wrong.

As for the test failure, I encountered the same failure in a different context while iterating on the PR, and it happened because the mock flake8 package provides blas, which is needed by the tested spec hypre, which depends on blas.  But flake8 wasn't meant to be actually used since it seems to have a bunch of fetch strategy attributes as a test that kind of thing isn't allowed.  Maybe we could fix that once and for all by removing the `provides('blas')` from the flake8 mock package?

scottwittenburg commented on lib/spack/spack/spec.py:
```diff
@@ -1649,11 +1649,24 @@ def node_dict_with_hashes(self, hash=ht.dag_hash):
+            # recompute full_hash or build_hash for old specs, and we
+            # only write these if they're present.
+            if self._full_hash:
+                node[self.name]['full_hash'] = self._full_hash
```
Sorry, I intended my [comment](https://github.com/spack/spack/pull/19672#pullrequestreview-521844640) to go here.

tgamblin committed:
bugfix: don't recompute full_hash or build_hash

There was an error introduced in #19209 where `full_hash()` and
`build_hash()` are called on older specs that we've read in from the DB.

Since we evaluate hashes lazily, `node_dict_with_hashes()` has no way of
knowing whether the spec just never had a `full_hash` or a `build_hash`
or whether they need to be lazily computed. When we just had `dag_hash`,
this wasn't an issue since it was always computed when concrete specs
were stored -- it would always be there for specs read in from files.

To get around this ambiguity and to fix the issue, we:

- [x] Add an attribute to the spec called `_hashes_final`, that is `True`
      if we can't lazily compute `build_hash` and `full_hash`.

- [x] Set `_hashes_final` to `False` for new specs (i.e., lazily
      computing hashes is ok)

- [x] Set `_hashes_final` to `True` for concrete specs read in via
      `from_node_dict`, as it may be too late to recompute hashes.

- [x] Compute and write out all hashes in `node_dict_with_hashes` *if
      possible*.

Effectively what this means is that we can round-trip specs that are
missing `_build_hash` and `_full_hash` without recomputing them, but for
all new specs, we'll compute them and store them. So Spack should work
fine with old DBs now.

This *may* make concretization slightly slower. If it does we could look
at some other way to store whether full_hash and build_hash can be
computed (e.g. we coudl store a `bool` on the `Spec` at concretize time),
or we could look at other options like using JSON for these hashes to
speed them up.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -1650,10 +1658,33 @@ def node_dict_with_hashes(self, hash=ht.dag_hash):
+                not self._hashes_final)                     # lazily compute
+            write_build_hash = 'build' in hash.deptype and (
+                self._hashes_final and self._build_hash or  # cached and final
+                not self.hashes_final)                      # lazily compute
```
Tests are failing because this needs to be `_hashes_final`
